fix(utxo): fix header deserialization; guard AuxPoW

shamardy opened the pull request:
Should close https://github.com/KomodoPlatform/komodo-defi-framework/issues/2048 when complete

shamardy committed:
serialization/reader: map RVN forks to CoinVariant::RVN (KawPoW headers)

shamardy committed:
stop false AuxPoW parsing on BTC-like forks; map NAV as BTC and add NAV regression test

shamardy committed:
treat RUNES (Runebase) as a Qtum fork to fix header deserialization

shamardy committed:
treat DOGEC as a PIVX fork

shamardy committed:
treat RIC as a BTC fork

shamardy committed:
treat AVN as a RVN fork

cipig reviewed (approved):
thanks for the fix
all coins from https://github.com/KomodoPlatform/komodo-defi-framework/issues/2048 now work fine
and tested almost 50 others, just to be sure

shamardy committed:
add context to header parse errors in electrum clients `get_median_time_past`

shamardy committed:
add coin/hex context to tx deserialize errors in `tx_enum_from_bytes`

- “UnexpectedEnd” without context can be misleading; including coin and the raw hex helps quickly pinpoint issues and not confuse it with a block header deserialization problem

shamardy committed:
include coin/hex (and hash when available) in tx deserialization errors

onur-ozkan commented:
Should this also resolve https://github.com/KomodoPlatform/komodo-defi-framework/issues/2584?

shamardy commented:
> Should this also resolve #2584?

No, this is only for block headers. It did add better logging to get more info such as the info we have here now https://github.com/KomodoPlatform/komodo-defi-framework/issues/2584#issue-3305963726

mariocynicys commented on mm2src/coins/utxo/rpc_clients/electrum_rpc/client.rs:
```diff
@@ -1103,18 +1103,28 @@ impl UtxoRpcClientOps for ElectrumClient {
-                    let headers = reader.read_list::<BlockHeader>()?;
+
+                    let headers = reader.read_list::<BlockHeader>().map_to_mm(|e| UtxoRpcError::InvalidResponse(format!(
+                            "blockchain.block.headers: failed to parse list of {} headers (coin={}, from={}, requested_count={}): {}",
```
im confused why clippy complains sometimes about format arguments not being inlined into the string and sometimes it's ok with it :/

shamardy replied:
Same here 😅

mariocynicys commented on mm2src/mm2_bitcoin/chain/src/block_header.rs:
```diff
@@ -270,7 +270,9 @@ impl Deserializable for BlockHeader {
         // https://en.bitcoin.it/wiki/Merged_mining_specification#Merged_mining_coinbase
-        let aux_pow = if (version & AUXPOW_VERSION_FLAG) != 0 {
+        // AuxPoW is only for AuxPoW coins (Dogecoin, Namecoin, Syscoin, etc.).
+        // BTC-like variants (BTC, BCH, NAV, …) must NOT interpret bit 8 as AuxPoW.
```
nitty nit: could we have an ascii 3 dots instead of this character (after `NAV, `). i find it a weird character xD

shamardy replied:
That's great attention to details

mariocynicys reviewed (dismissed):
LGTM!
Thanks

shamardy committed:
fix 3 dots nit
